{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/AnastasiaBrinati/Progetto-ML-23-24/blob/main/task1_federato.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import tensorflow as tf\n",
        "\n",
        "!pip install -q flwr[\"simulation\"] tensorflow\n",
        "!pip install -q flwr_datasets[\"vision\"]\n",
        "!pip install matplotlib\n",
        "\n",
        "from typing import Dict, List, Tuple\n",
        "import flwr as fl\n",
        "from flwr.common import Metrics\n",
        "import tensorflow as tf\n",
        "from tensorflow.keras.initializers import he_normal, glorot_normal\n",
        "from tensorflow.keras.layers import Input, Dense, concatenate, Flatten\n",
        "!pip install tensorflow-addons\n",
        "from tensorflow_addons.metrics import F1Score\n",
        "from flwr.simulation.ray_transport.utils import enable_tf_gpu_growth\n",
        "from datasets import Dataset\n",
        "from flwr_datasets import FederatedDataset\n",
        "from keras.initializers import RandomNormal, he_normal, glorot_normal, he_uniform, glorot_uniform"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c1tJ8LG1hr7Z",
        "outputId": "48ebb485-598a-4d02-ff78-3fc28daf4516"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (3.7.1)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (4.49.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.4.5)\n",
            "Requirement already satisfied: numpy>=1.20 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.25.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (23.2)\n",
            "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (9.4.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (3.1.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
            "Requirement already satisfied: tensorflow-addons in /usr/local/lib/python3.10/dist-packages (0.23.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from tensorflow-addons) (23.2)\n",
            "Requirement already satisfied: typeguard<3.0.0,>=2.7 in /usr/local/lib/python3.10/dist-packages (from tensorflow-addons) (2.13.3)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "wr1W-6_43eIE"
      },
      "outputs": [],
      "source": [
        "# Simulation parameters\n",
        "settings = {\n",
        "    \"VERBOSE\": 0,\n",
        "    \"NUM_CLIENTS\": 100,\n",
        "    \"FRACTION_FIT\": 0.1,\n",
        "    \"FRACTION_EVALUATE\": 0.05,\n",
        "    \"MIN_FIT_CLIENTS\": 10,\n",
        "    \"MIN_EVALUATE_CLIENTS\": 5,\n",
        "    \"MIN_AVAILABLE_CLIENTS_FRACTION\": 0.75,\n",
        "    \"STRATEGY_TYPE\": 'FedOptim', # FedOptim, FedAvg, FedProx\n",
        "    \"NUM_ROUNDS\": 10\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "Z778yKgU3eIF"
      },
      "outputs": [],
      "source": [
        "\n",
        "def get_model():\n",
        "    \"\"\"Constructs a model suitable for features.\"\"\"\n",
        "    # Define input layers\n",
        "    Support_Calls_input = Input(shape=(1,), name='Support Calls')\n",
        "    Payment_Delay_input = Input(shape=(1,), name='Payment Delay')\n",
        "    Usage_Frequency_input = Input(shape=(1,), name='Usage Frequency')\n",
        "    Total_Spend_input = Input(shape=(1,), name='Total Spend')\n",
        "    Last_Interaction_input = Input(shape=(1,), name='Last Interaction')\n",
        "    Age_18_25_input = Input(shape=(1,), name='Age_binned_(18.0, 25.0]')\n",
        "    Age_25_35_input = Input(shape=(1,), name='Age_binned_(25.0, 35.0]')\n",
        "    Age_35_50_input = Input(shape=(1,), name='Age_binned_(35.0, 50.0]')\n",
        "    Age_50_60_input = Input(shape=(1,), name='Age_binned_(50.0, 60.0]')\n",
        "    Age_60_70_input = Input(shape=(1,), name='Age_binned_(60.0, 70.0]')\n",
        "    Age_nan_input = Input(shape=(1,), name='Age_binned_nan')\n",
        "    Contract_Length_Annual_input = Input(shape=(1,), name='Contract Length_Annual')\n",
        "    Contract_Length_Monthly_input = Input(shape=(1,), name='Contract Length_Monthly')\n",
        "    Contract_Length_Quarterly_input = Input(shape=(1,), name='Contract Length_Quarterly')\n",
        "    Contract_Length_nan_input = Input(shape=(1,), name='Contract Length_nan')\n",
        "    Gender_Female_input = Input(shape=(1,), name='Gender_Female')\n",
        "    Gender_Male_input = Input(shape=(1,), name='Gender_Male')\n",
        "    Gender_nan_input = Input(shape=(1,), name='Gender_nan')\n",
        "    Subscription_Type_Basic_input = Input(shape=(1,), name='Subscription Type_Basic')\n",
        "    Subscription_Type_Premium_input = Input(shape=(1,), name='Subscription Type_Premium')\n",
        "    Subscription_Type_Standard_input = Input(shape=(1,), name='Subscription Type_Standard')\n",
        "    Subscription_Type_nan_input = Input(shape=(1,), name='Subscription Type_nan')\n",
        "\n",
        "    # Concatenate the input tensors\n",
        "    concatenated_inputs = concatenate([\n",
        "        Support_Calls_input, Payment_Delay_input, Usage_Frequency_input, Total_Spend_input, Last_Interaction_input,\n",
        "        Age_18_25_input, Age_25_35_input, Age_35_50_input, Age_50_60_input, Age_60_70_input, Age_nan_input,\n",
        "        Contract_Length_Annual_input, Contract_Length_Monthly_input, Contract_Length_Quarterly_input, Contract_Length_nan_input,\n",
        "        Gender_Female_input, Gender_Male_input, Gender_nan_input,\n",
        "        Subscription_Type_Basic_input, Subscription_Type_Premium_input, Subscription_Type_Standard_input, Subscription_Type_nan_input\n",
        "    ])\n",
        "\n",
        "    # Define the rest of the model\n",
        "    x = Flatten()(concatenated_inputs)\n",
        "    x = Dense(80, activation=\"PReLU\", kernel_initializer=he_normal())(x)\n",
        "    x = Dense(64, activation=\"relu\", kernel_initializer=he_normal())(x)\n",
        "    x = Dense(16, activation=\"tanh\", kernel_initializer=he_normal())(x)\n",
        "    x = Dense(32, activation=\"sigmoid\", kernel_initializer=glorot_normal())(x)\n",
        "    output = Dense(1, activation=\"sigmoid\", kernel_initializer=glorot_normal())(x)\n",
        "\n",
        "    # Construct the model\n",
        "    model = tf.keras.models.Model(inputs=[\n",
        "        Support_Calls_input, Payment_Delay_input, Usage_Frequency_input, Total_Spend_input, Last_Interaction_input,\n",
        "        Age_18_25_input, Age_25_35_input, Age_35_50_input, Age_50_60_input, Age_60_70_input, Age_nan_input,\n",
        "        Contract_Length_Annual_input, Contract_Length_Monthly_input, Contract_Length_Quarterly_input, Contract_Length_nan_input,\n",
        "        Gender_Female_input, Gender_Male_input, Gender_nan_input,\n",
        "        Subscription_Type_Basic_input, Subscription_Type_Premium_input, Subscription_Type_Standard_input, Subscription_Type_nan_input\n",
        "    ], outputs=output)\n",
        "\n",
        "    # Compile the model with accuracy, precision, recall, and F1-score metrics\n",
        "    model.compile(optimizer=\"adam\", loss=\"binary_crossentropy\", metrics=[\"accuracy\", tf.keras.metrics.Precision(), tf.keras.metrics.Recall(),tf.keras.metrics.F1Score()])\n",
        "    return model\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "CreaAmfK3eIG"
      },
      "outputs": [],
      "source": [
        "class FlowerClient(fl.client.NumPyClient):\n",
        "    def __init__(self, trainset, valset) -> None:\n",
        "        # Create model\n",
        "        self.model = get_model()\n",
        "        self.trainset = trainset\n",
        "        self.valset = valset\n",
        "\n",
        "    def get_parameters(self, config):\n",
        "        return self.model.get_weights()\n",
        "\n",
        "    def fit(self, parameters, config):\n",
        "        self.model.set_weights(parameters)\n",
        "        self.model.fit(self.trainset, epochs=1, verbose=settings['VERBOSE'])\n",
        "        return self.model.get_weights(), len(self.trainset), {}\n",
        "\n",
        "    def evaluate(self, parameters, config):\n",
        "        self.model.set_weights(parameters)\n",
        "        results = self.model.evaluate(self.valset)\n",
        "\n",
        "        loss = results[0]\n",
        "        accuracy = results[1]\n",
        "        precision = results[2]\n",
        "        recall = results[3]\n",
        "        f1_score = results[4]\n",
        "\n",
        "        # Convert metrics results to appropriate data types\n",
        "        loss = float(loss)\n",
        "        accuracy = float(accuracy)\n",
        "        precision = float(precision)\n",
        "        recall = float(recall)\n",
        "        f1_score = float(f1_score)\n",
        "\n",
        "\n",
        "        return loss, len(self.valset), {\"accuracy\": accuracy, \"precision\":precision, \"recall\":recall, \"f1_score\":f1_score}"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download MNIST dataset and partition it\n",
        "mnist_fds = FederatedDataset(dataset=\"giulioappetito/churn_dataset_giulioappetito\", partitioners={\"train\": settings['NUM_CLIENTS']})\n",
        "# Get the whole test set for centralised evaluation\n",
        "centralized_testset = mnist_fds.load_full(\"test\").to_tf_dataset(\n",
        "  columns=[\n",
        "            'Support Calls',\n",
        "            'Payment Delay',\n",
        "            'Usage Frequency',\n",
        "            'Total Spend',\n",
        "            'Last Interaction',\n",
        "            'Age_binned_(18.0, 25.0]',\n",
        "            'Age_binned_(25.0, 35.0]',\n",
        "            'Age_binned_(35.0, 50.0]',\n",
        "            'Age_binned_(50.0, 60.0]',\n",
        "            'Age_binned_(60.0, 70.0]',\n",
        "            'Age_binned_nan',\n",
        "            'Contract Length_Annual',\n",
        "            'Contract Length_Monthly',\n",
        "            'Contract Length_Quarterly',\n",
        "            'Contract Length_nan',\n",
        "            'Gender_Female',\n",
        "            'Gender_Male',\n",
        "            'Gender_nan',\n",
        "            'Subscription Type_Basic',\n",
        "            'Subscription Type_Premium',\n",
        "            'Subscription Type_Standard',\n",
        "            'Subscription Type_nan'\n",
        "        ],\n",
        "        label_cols=\"Churn\", batch_size=64\n",
        "    )"
      ],
      "metadata": {
        "id": "b3QVqYkdWxQw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6147f82c-3e6c-4a8f-8252-46d5112edb3c"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/flwr_datasets/utils.py:84: UserWarning: The currently tested dataset are ['mnist', 'cifar10', 'fashion_mnist', 'sasha/dog-food', 'zh-plus/tiny-imagenet']. Given: giulioappetito/churn_dataset_giulioappetito.\n",
            "  warnings.warn(\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "BJYjw7yt3eIH"
      },
      "outputs": [],
      "source": [
        "# @title\n",
        "def get_client_fn(dataset: FederatedDataset):\n",
        "    \"\"\"Return a function to construct a client.\n",
        "\n",
        "    The VirtualClientEngine will execute this function whenever a client is sampled by\n",
        "    the strategy to participate.\n",
        "    \"\"\"\n",
        "\n",
        "    def client_fn(cid: str) -> fl.client.Client:\n",
        "        \"\"\"Construct a FlowerClient with its own dataset partition.\"\"\"\n",
        "\n",
        "        # Extract partition for client with id = cid\n",
        "        client_dataset = dataset.load_partition(int(cid), \"train\")\n",
        "\n",
        "        # Now let's split it into train (90%) and validation (10%)\n",
        "        client_dataset_splits = client_dataset.train_test_split(test_size=0.1)\n",
        "        trainset = client_dataset_splits[\"train\"].to_tf_dataset(\n",
        "            columns=[\n",
        "                'Support Calls',\n",
        "                'Payment Delay',\n",
        "                'Usage Frequency',\n",
        "                'Total Spend',\n",
        "                'Last Interaction',\n",
        "                'Age_binned_(18.0, 25.0]',\n",
        "                'Age_binned_(25.0, 35.0]',\n",
        "                'Age_binned_(35.0, 50.0]',\n",
        "                'Age_binned_(50.0, 60.0]',\n",
        "                'Age_binned_(60.0, 70.0]',\n",
        "                'Age_binned_nan',\n",
        "                'Contract Length_Annual',\n",
        "                'Contract Length_Monthly',\n",
        "                'Contract Length_Quarterly',\n",
        "                'Contract Length_nan',\n",
        "                'Gender_Female',\n",
        "                'Gender_Male',\n",
        "                'Gender_nan',\n",
        "                'Subscription Type_Basic',\n",
        "                'Subscription Type_Premium',\n",
        "                'Subscription Type_Standard',\n",
        "                'Subscription Type_nan'\n",
        "            ],\n",
        "            label_cols=\"Churn\", batch_size=32\n",
        "        )\n",
        "        valset = client_dataset_splits[\"test\"].to_tf_dataset(\n",
        "            columns=[\n",
        "                'Support Calls',\n",
        "                'Payment Delay',\n",
        "                'Usage Frequency',\n",
        "                'Total Spend',\n",
        "                'Last Interaction',\n",
        "                'Age_binned_(18.0, 25.0]',\n",
        "                'Age_binned_(25.0, 35.0]',\n",
        "                'Age_binned_(35.0, 50.0]',\n",
        "                'Age_binned_(50.0, 60.0]',\n",
        "                'Age_binned_(60.0, 70.0]',\n",
        "                'Age_binned_nan',\n",
        "                'Contract Length_Annual',\n",
        "                'Contract Length_Monthly',\n",
        "                'Contract Length_Quarterly',\n",
        "                'Contract Length_nan',\n",
        "                'Gender_Female',\n",
        "                'Gender_Male',\n",
        "                'Gender_nan',\n",
        "                'Subscription Type_Basic',\n",
        "                'Subscription Type_Premium',\n",
        "                'Subscription Type_Standard',\n",
        "                'Subscription Type_nan'\n",
        "            ],\n",
        "            label_cols=\"Churn\", batch_size=64\n",
        "        )\n",
        "\n",
        "        # Create and return client\n",
        "        return FlowerClient(trainset, valset).to_client()\n",
        "\n",
        "    return client_fn\n",
        "\n",
        "\n",
        "def weighted_average(metrics: List[Tuple[int, Metrics]]) -> Metrics:\n",
        "    \"\"\"Aggregation function for (federated) evaluation metrics, i.e. those returned by\n",
        "    the client's evaluate() method.\"\"\"\n",
        "    # Multiply accuracy of each client by number of examples used\n",
        "    accuracies = [num_examples * m[\"accuracy\"] for num_examples, m in metrics]\n",
        "    precisions = [num_examples * m[\"precision\"] for num_examples, m in metrics]\n",
        "    recalls = [num_examples * m[\"recall\"] for num_examples, m in metrics]\n",
        "    f1_scores = [num_examples * m[\"f1_score\"] for num_examples, m in metrics]\n",
        "    examples = [num_examples for num_examples, _ in metrics]\n",
        "\n",
        "    # Aggregate and return custom metric (weighted average)\n",
        "    return {\"accuracy\": sum(accuracies) / sum(examples), \"precision\":sum(precisions) / sum(examples),\"recall\":sum(recalls) / sum(examples),\"f1_score\":sum(f1_scores) / sum(examples)}\n",
        "\n",
        "\n",
        "def get_evaluate_fn(testset: Dataset):\n",
        "    \"\"\"Return an evaluation function for server-side (i.e. centralised) evaluation.\"\"\"\n",
        "\n",
        "    # The `evaluate` function will be called after every round by the strategy\n",
        "    def evaluate(\n",
        "        server_round: int,\n",
        "        parameters: fl.common.NDArrays,\n",
        "        config: Dict[str, fl.common.Scalar],\n",
        "    ):\n",
        "        model = get_model()  # Construct the model\n",
        "        model.set_weights(parameters)  # Update model with the latest parameters\n",
        "        results = model.evaluate(testset)\n",
        "        loss = results[0]\n",
        "        return loss, {\"accuracy\": results[1], \"precision\":results[2], \"recall\":results[3], \"f1_score\":results[4]}\n",
        "\n",
        "    return evaluate\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Create FedAvg strategy\n",
        "strategy_FedOptim = fl.server.strategy.FedProx(\n",
        "    proximal_mu=1.0,\n",
        "    fraction_fit=settings[\"FRACTION_FIT\"],\n",
        "    fraction_evaluate=settings[\"FRACTION_EVALUATE\"],\n",
        "    min_fit_clients=settings[\"MIN_FIT_CLIENTS\"],\n",
        "    min_evaluate_clients=settings[\"MIN_EVALUATE_CLIENTS\"],\n",
        "    min_available_clients=int(settings[\"NUM_CLIENTS\"] * settings[\"MIN_AVAILABLE_CLIENTS_FRACTION\"]),\n",
        "    evaluate_metrics_aggregation_fn=weighted_average,\n",
        "    evaluate_fn=get_evaluate_fn(centralized_testset),\n",
        ")\n",
        "\n",
        "strategy_FedProx = fl.server.strategy.FedProx(\n",
        "    proximal_mu=1.0,\n",
        "    fraction_fit=settings[\"FRACTION_FIT\"],\n",
        "    fraction_evaluate=settings[\"FRACTION_EVALUATE\"],\n",
        "    min_fit_clients=settings[\"MIN_FIT_CLIENTS\"],\n",
        "    min_evaluate_clients=settings[\"MIN_EVALUATE_CLIENTS\"],\n",
        "    min_available_clients=int(settings[\"NUM_CLIENTS\"] * settings[\"MIN_AVAILABLE_CLIENTS_FRACTION\"]),\n",
        "    evaluate_metrics_aggregation_fn=weighted_average,\n",
        "    evaluate_fn=get_evaluate_fn(centralized_testset),\n",
        ")\n",
        "\n",
        "strategy_FedAvg = fl.server.strategy.FedAvg(\n",
        "    fraction_fit=settings[\"FRACTION_FIT\"],\n",
        "    fraction_evaluate=settings[\"FRACTION_EVALUATE\"],\n",
        "    min_fit_clients=settings[\"MIN_FIT_CLIENTS\"],\n",
        "    min_evaluate_clients=settings[\"MIN_EVALUATE_CLIENTS\"],\n",
        "    min_available_clients=int(settings[\"NUM_CLIENTS\"] * settings[\"MIN_AVAILABLE_CLIENTS_FRACTION\"]),\n",
        "    evaluate_metrics_aggregation_fn=weighted_average,\n",
        "    evaluate_fn=get_evaluate_fn(centralized_testset),\n",
        ")\n",
        "\n",
        "# Selezione della strategia\n",
        "if settings[\"STRATEGY_TYPE\"] == 'FedProx':\n",
        "    strategy = strategy_FedProx\n",
        "elif settings[\"STRATEGY_TYPE\"] == 'FedOptim':\n",
        "    strategy = strategy_FedOptim\n",
        "else:\n",
        "    strategy = strategy_FedAvg\n",
        "\n",
        "# Configurazione delle risorse del client\n",
        "client_resources = {\"num_cpus\": 1, \"num_gpus\": 0.0}\n",
        "\n",
        "# Avvia la simulazione una volta\n",
        "history = fl.simulation.start_simulation(\n",
        "    client_fn=get_client_fn(mnist_fds),\n",
        "    num_clients=settings[\"NUM_CLIENTS\"],\n",
        "    config=fl.server.ServerConfig(num_rounds=settings[\"NUM_ROUNDS\"]),\n",
        "    strategy=strategy,\n",
        "    client_resources=client_resources,\n",
        "    actor_kwargs={\"on_actor_init_fn\": enable_tf_gpu_growth}  # Abilita la crescita della GPU durante l'inizializzazione dell'attore\n",
        ")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "6M_HYuC67-6_",
        "outputId": "9cabe1aa-0c6d-48b2-f9a2-73c31d650f62"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "INFO flwr 2024-02-29 13:19:02,587 | app.py:178 | Starting Flower simulation, config: ServerConfig(num_rounds=10, round_timeout=None)\n",
            "INFO:flwr:Starting Flower simulation, config: ServerConfig(num_rounds=10, round_timeout=None)\n",
            "2024-02-29 13:19:12,959\tINFO worker.py:1621 -- Started a local Ray instance.\n",
            "INFO flwr 2024-02-29 13:19:17,878 | app.py:213 | Flower VCE: Ray initialized with resources: {'node:__internal_head__': 1.0, 'node:172.28.0.12': 1.0, 'CPU': 2.0, 'memory': 7842462107.0, 'object_store_memory': 3921231052.0}\n",
            "INFO:flwr:Flower VCE: Ray initialized with resources: {'node:__internal_head__': 1.0, 'node:172.28.0.12': 1.0, 'CPU': 2.0, 'memory': 7842462107.0, 'object_store_memory': 3921231052.0}\n",
            "INFO flwr 2024-02-29 13:19:17,897 | app.py:219 | Optimize your simulation with Flower VCE: https://flower.dev/docs/framework/how-to-run-simulations.html\n",
            "INFO:flwr:Optimize your simulation with Flower VCE: https://flower.dev/docs/framework/how-to-run-simulations.html\n",
            "INFO flwr 2024-02-29 13:19:17,903 | app.py:242 | Flower VCE: Resources for each Virtual Client: {'num_cpus': 1, 'num_gpus': 0.0}\n",
            "INFO:flwr:Flower VCE: Resources for each Virtual Client: {'num_cpus': 1, 'num_gpus': 0.0}\n",
            "INFO flwr 2024-02-29 13:19:18,015 | app.py:288 | Flower VCE: Creating VirtualClientEngineActorPool with 2 actors\n",
            "INFO:flwr:Flower VCE: Creating VirtualClientEngineActorPool with 2 actors\n",
            "INFO flwr 2024-02-29 13:19:18,025 | server.py:89 | Initializing global parameters\n",
            "INFO:flwr:Initializing global parameters\n",
            "INFO flwr 2024-02-29 13:19:18,028 | server.py:276 | Requesting initial parameters from one random client\n",
            "INFO:flwr:Requesting initial parameters from one random client\n",
            "\u001b[2m\u001b[36m(pid=8828)\u001b[0m 2024-02-29 13:19:28.384068: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "\u001b[2m\u001b[36m(pid=8828)\u001b[0m 2024-02-29 13:19:28.386193: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "\u001b[2m\u001b[36m(pid=8828)\u001b[0m 2024-02-29 13:19:28.390708: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\u001b[2m\u001b[36m(pid=8830)\u001b[0m 2024-02-29 13:19:38.016686: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
            "\u001b[2m\u001b[36m(pid=8830)\u001b[0m 2024-02-29 13:19:28.414585: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "\u001b[2m\u001b[36m(pid=8830)\u001b[0m 2024-02-29 13:19:28.414679: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "\u001b[2m\u001b[36m(pid=8830)\u001b[0m 2024-02-29 13:19:28.419330: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "INFO flwr 2024-02-29 13:19:46,247 | server.py:280 | Received initial parameters from one random client\n",
            "INFO:flwr:Received initial parameters from one random client\n",
            "INFO flwr 2024-02-29 13:19:46,253 | server.py:91 | Evaluating initial parameters\n",
            "INFO:flwr:Evaluating initial parameters\n",
            "ERROR flwr 2024-02-29 13:19:46,830 | app.py:313 | in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2066, in test_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2049, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2037, in run_step  **\n",
            "        outputs = model.test_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1920, in test_step\n",
            "        return self.compute_metrics(x, y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1249, in compute_metrics\n",
            "        self.compiled_metrics.update_state(y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 620, in update_state\n",
            "        metric_obj.update_state(y_t, y_p, sample_weight=mask)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/metrics_utils.py\", line 77, in decorated\n",
            "        result = update_state_fn(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/base_metric.py\", line 140, in update_state_fn\n",
            "        return ag_update_state(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/f_score_metrics.py\", line 176, in update_state  **\n",
            "        y_true = tf.convert_to_tensor(y_true, dtype=self.dtype)\n",
            "\n",
            "    ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int64: <tf.Tensor 'ExpandDims_1:0' shape=(None, 1) dtype=int64>\n",
            "\n",
            "ERROR:flwr:in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2066, in test_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2049, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2037, in run_step  **\n",
            "        outputs = model.test_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1920, in test_step\n",
            "        return self.compute_metrics(x, y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1249, in compute_metrics\n",
            "        self.compiled_metrics.update_state(y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 620, in update_state\n",
            "        metric_obj.update_state(y_t, y_p, sample_weight=mask)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/metrics_utils.py\", line 77, in decorated\n",
            "        result = update_state_fn(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/base_metric.py\", line 140, in update_state_fn\n",
            "        return ag_update_state(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/f_score_metrics.py\", line 176, in update_state  **\n",
            "        y_true = tf.convert_to_tensor(y_true, dtype=self.dtype)\n",
            "\n",
            "    ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int64: <tf.Tensor 'ExpandDims_1:0' shape=(None, 1) dtype=int64>\n",
            "\n",
            "ERROR flwr 2024-02-29 13:19:46,837 | app.py:314 | Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/simulation/app.py\", line 308, in start_simulation\n",
            "    hist = run_fl(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/app.py\", line 225, in run_fl\n",
            "    hist = server.fit(num_rounds=config.num_rounds, timeout=config.round_timeout)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/server.py\", line 92, in fit\n",
            "    res = self.strategy.evaluate(0, parameters=self.parameters)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/strategy/fedavg.py\", line 165, in evaluate\n",
            "    eval_res = self.evaluate_fn(server_round, parameters_ndarrays, {})\n",
            "  File \"<ipython-input-13-2ceed5580f1a>\", line 103, in evaluate\n",
            "    results = model.evaluate(testset)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "    raise e.with_traceback(filtered_tb) from None\n",
            "  File \"/tmp/__autograph_generated_filemd00o7qd.py\", line 15, in tf__test_function\n",
            "    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)\n",
            "ValueError: in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2066, in test_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2049, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2037, in run_step  **\n",
            "        outputs = model.test_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1920, in test_step\n",
            "        return self.compute_metrics(x, y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1249, in compute_metrics\n",
            "        self.compiled_metrics.update_state(y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 620, in update_state\n",
            "        metric_obj.update_state(y_t, y_p, sample_weight=mask)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/metrics_utils.py\", line 77, in decorated\n",
            "        result = update_state_fn(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/base_metric.py\", line 140, in update_state_fn\n",
            "        return ag_update_state(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/f_score_metrics.py\", line 176, in update_state  **\n",
            "        y_true = tf.convert_to_tensor(y_true, dtype=self.dtype)\n",
            "\n",
            "    ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int64: <tf.Tensor 'ExpandDims_1:0' shape=(None, 1) dtype=int64>\n",
            "\n",
            "\n",
            "ERROR:flwr:Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/simulation/app.py\", line 308, in start_simulation\n",
            "    hist = run_fl(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/app.py\", line 225, in run_fl\n",
            "    hist = server.fit(num_rounds=config.num_rounds, timeout=config.round_timeout)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/server.py\", line 92, in fit\n",
            "    res = self.strategy.evaluate(0, parameters=self.parameters)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/flwr/server/strategy/fedavg.py\", line 165, in evaluate\n",
            "    eval_res = self.evaluate_fn(server_round, parameters_ndarrays, {})\n",
            "  File \"<ipython-input-13-2ceed5580f1a>\", line 103, in evaluate\n",
            "    results = model.evaluate(testset)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
            "    raise e.with_traceback(filtered_tb) from None\n",
            "  File \"/tmp/__autograph_generated_filemd00o7qd.py\", line 15, in tf__test_function\n",
            "    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)\n",
            "ValueError: in user code:\n",
            "\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2066, in test_function  *\n",
            "        return step_function(self, iterator)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2049, in step_function  **\n",
            "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2037, in run_step  **\n",
            "        outputs = model.test_step(data)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1920, in test_step\n",
            "        return self.compute_metrics(x, y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1249, in compute_metrics\n",
            "        self.compiled_metrics.update_state(y, y_pred, sample_weight)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 620, in update_state\n",
            "        metric_obj.update_state(y_t, y_p, sample_weight=mask)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/metrics_utils.py\", line 77, in decorated\n",
            "        result = update_state_fn(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/base_metric.py\", line 140, in update_state_fn\n",
            "        return ag_update_state(*args, **kwargs)\n",
            "    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/f_score_metrics.py\", line 176, in update_state  **\n",
            "        y_true = tf.convert_to_tensor(y_true, dtype=self.dtype)\n",
            "\n",
            "    ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int64: <tf.Tensor 'ExpandDims_1:0' shape=(None, 1) dtype=int64>\n",
            "\n",
            "\n",
            "ERROR flwr 2024-02-29 13:19:46,843 | app.py:315 | Your simulation crashed :(. This could be because of several reasons. The most common are: \n",
            "\t > Sometimes, issues in the simulation code itself can cause crashes. It's always a good idea to double-check your code for any potential bugs or inconsistencies that might be contributing to the problem. For example: \n",
            "\t\t - You might be using a class attribute in your clients that hasn't been defined.\n",
            "\t\t - There could be an incorrect method call to a 3rd party library (e.g., PyTorch).\n",
            "\t\t - The return types of methods in your clients/strategies might be incorrect.\n",
            "\t > Your system couldn't fit a single VirtualClient: try lowering `client_resources`.\n",
            "\t > All the actors in your pool crashed. This could be because: \n",
            "\t\t - You clients hit an out-of-memory (OOM) error and actors couldn't recover from it. Try launching your simulation with more generous `client_resources` setting (i.e. it seems {'num_cpus': 1, 'num_gpus': 0.0} is not enough for your run). Use fewer concurrent actors. \n",
            "\t\t - You were running a multi-node simulation and all worker nodes disconnected. The head node might still be alive but cannot accommodate any actor with resources: {'num_cpus': 1, 'num_gpus': 0.0}.\n",
            "Take a look at the Flower simulation examples for guidance <https://flower.dev/docs/framework/how-to-run-simulations.html>.\n",
            "ERROR:flwr:Your simulation crashed :(. This could be because of several reasons. The most common are: \n",
            "\t > Sometimes, issues in the simulation code itself can cause crashes. It's always a good idea to double-check your code for any potential bugs or inconsistencies that might be contributing to the problem. For example: \n",
            "\t\t - You might be using a class attribute in your clients that hasn't been defined.\n",
            "\t\t - There could be an incorrect method call to a 3rd party library (e.g., PyTorch).\n",
            "\t\t - The return types of methods in your clients/strategies might be incorrect.\n",
            "\t > Your system couldn't fit a single VirtualClient: try lowering `client_resources`.\n",
            "\t > All the actors in your pool crashed. This could be because: \n",
            "\t\t - You clients hit an out-of-memory (OOM) error and actors couldn't recover from it. Try launching your simulation with more generous `client_resources` setting (i.e. it seems {'num_cpus': 1, 'num_gpus': 0.0} is not enough for your run). Use fewer concurrent actors. \n",
            "\t\t - You were running a multi-node simulation and all worker nodes disconnected. The head node might still be alive but cannot accommodate any actor with resources: {'num_cpus': 1, 'num_gpus': 0.0}.\n",
            "Take a look at the Flower simulation examples for guidance <https://flower.dev/docs/framework/how-to-run-simulations.html>.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "RuntimeError",
          "evalue": "Simulation crashed.",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/flwr/simulation/app.py\u001b[0m in \u001b[0;36mstart_simulation\u001b[0;34m(client_fn, num_clients, clients_ids, client_resources, server, config, strategy, client_manager, ray_init_args, keep_initialised, actor_type, actor_kwargs, actor_scheduling)\u001b[0m\n\u001b[1;32m    307\u001b[0m         \u001b[0;31m# Start training\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 308\u001b[0;31m         hist = run_fl(\n\u001b[0m\u001b[1;32m    309\u001b[0m             \u001b[0mserver\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minitialized_server\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/flwr/server/app.py\u001b[0m in \u001b[0;36mrun_fl\u001b[0;34m(server, config)\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0;34m\"\"\"Train a model on the given server and return the History object.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m     \u001b[0mhist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum_rounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnum_rounds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mround_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m     \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"app_fit: losses_distributed %s\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlosses_distributed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/flwr/server/server.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, num_rounds, timeout)\u001b[0m\n\u001b[1;32m     91\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mINFO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Evaluating initial parameters\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 92\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrategy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevaluate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/flwr/server/strategy/fedavg.py\u001b[0m in \u001b[0;36mevaluate\u001b[0;34m(self, server_round, parameters)\u001b[0m\n\u001b[1;32m    164\u001b[0m         \u001b[0mparameters_ndarrays\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparameters_to_ndarrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 165\u001b[0;31m         \u001b[0meval_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevaluate_fn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mserver_round\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters_ndarrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    166\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0meval_res\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-13-2ceed5580f1a>\u001b[0m in \u001b[0;36mevaluate\u001b[0;34m(server_round, parameters, config)\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_weights\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Update model with the latest parameters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 103\u001b[0;31m         \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevaluate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtestset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    104\u001b[0m         \u001b[0mloss\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     69\u001b[0m             \u001b[0;31m# `tf.debugging.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\u001b[0m in \u001b[0;36mtf__test_function\u001b[0;34m(iterator)\u001b[0m\n\u001b[1;32m     14\u001b[0m                     \u001b[0mdo_return\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m                     \u001b[0mretval_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconverted_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstep_function\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfscope\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m                 \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: in user code:\n\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2066, in test_function  *\n        return step_function(self, iterator)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2049, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 2037, in run_step  **\n        outputs = model.test_step(data)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1920, in test_step\n        return self.compute_metrics(x, y, y_pred, sample_weight)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1249, in compute_metrics\n        self.compiled_metrics.update_state(y, y_pred, sample_weight)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 620, in update_state\n        metric_obj.update_state(y_t, y_p, sample_weight=mask)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/metrics_utils.py\", line 77, in decorated\n        result = update_state_fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/base_metric.py\", line 140, in update_state_fn\n        return ag_update_state(*args, **kwargs)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/metrics/f_score_metrics.py\", line 176, in update_state  **\n        y_true = tf.convert_to_tensor(y_true, dtype=self.dtype)\n\n    ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int64: <tf.Tensor 'ExpandDims_1:0' shape=(None, 1) dtype=int64>\n",
            "\nThe above exception was the direct cause of the following exception:\n",
            "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-14-50eea8a38b6e>\u001b[0m in \u001b[0;36m<cell line: 46>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;31m# Avvia la simulazione una volta\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m history = fl.simulation.start_simulation(\n\u001b[0m\u001b[1;32m     47\u001b[0m     \u001b[0mclient_fn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mget_client_fn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmnist_fds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m     \u001b[0mnum_clients\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"NUM_CLIENTS\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/flwr/simulation/app.py\u001b[0m in \u001b[0;36mstart_simulation\u001b[0;34m(client_fn, num_clients, clients_ids, client_resources, server, config, strategy, client_manager, ray_init_args, keep_initialised, actor_type, actor_kwargs, actor_scheduling)\u001b[0m\n\u001b[1;32m    342\u001b[0m             \u001b[0mclient_resources\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    343\u001b[0m         )\n\u001b[0;32m--> 344\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Simulation crashed.\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    345\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    346\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mRuntimeError\u001b[0m: Simulation crashed."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fa3sHQYT3eII"
      },
      "outputs": [],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from tabulate import tabulate\n",
        "\n",
        "# I dati dell'accuratezza\n",
        "print(f\"{history.metrics_centralized = }\")\n",
        "global_accuracy_centralised = history.metrics_centralized[\"accuracy\"]\n",
        "global_recall_centralised = history.metrics_centralized[\"recall\"]\n",
        "global_precision_centralised = history.metrics_centralized[\"precision\"]\n",
        "global_f1_score_centralised = history.metrics_centralized[\"f1_score\"]\n",
        "\n",
        "\n",
        "\n",
        "rounds = [data[0] for data in global_accuracy_centralised]\n",
        "accuracies = [100.0 * data[1] for data in global_accuracy_centralised]\n",
        "recalls = [100.0 * data[1] for data in global_recall_centralised]\n",
        "precisions = [100.0 * data[1] for data in global_precision_centralised]\n",
        "f1_scores = [100.0 * data[1] for data in global_f1_score_centralised]\n",
        "\n",
        "\n",
        "# Plot\n",
        "plt.plot(rounds, accuracies)\n",
        "plt.grid()\n",
        "plt.ylabel(\"Accuracy (%)\")\n",
        "plt.xlabel(\"Round\")\n",
        "plt.scatter(rounds,accuracies)\n",
        "\n",
        "plt.plot(rounds,recalls, label=\"Recall\")\n",
        "plt.scatter(rounds,recalls)\n",
        "\n",
        "plt.plot(rounds,precisions, label = \"Precision\")\n",
        "plt.scatter(rounds,precisions)\n",
        "\n",
        "plt.plot(rounds,f1_scores, label = \"F1 Score\")\n",
        "plt.scatter(rounds,f1_scores)\n",
        "\n",
        "\n",
        "plt.title(str(settings['NUM_CLIENTS']) + ' clients with ' + str(int(settings['FRACTION_FIT']*settings['NUM_CLIENTS'])) + ' clients per round, with strategy ' + settings['STRATEGY_TYPE'])  # Correzione qui\n",
        "plt.show()\n",
        "plt.legend()\n",
        "\n",
        "# Tabella\n",
        "data_table = [[\"Round\", \"Accuracy (%)\",\"Precision\",\"Recall\",\"F1_score\"]] + [[r, a,p,rec,f1] for r, a,p, rec,f1 in zip(rounds, accuracies, precisions, recalls, f1_scores)]\n",
        "table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\")\n",
        "print(table)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}